hoomd/logging.py

```diff
@@ -1,74 +1,252 @@
+
+    Flags:
+        scalar: `float` or `int` objects (i.e. numbers)
+
+        sequence: sequence (e.g. `list`, `tuple`, `numpy.ndarray`) of numbers of
+        the same type.
+
+        string: a single Python `str` object
+
+        strings: a sequence of Python `str` objects
+
+        object: any Python object outside a sequence, string, or scalar.
+
+        angle: per-angle quantity
+
+        bond: per-bond quantity
+
+        constraint: per-constraint quantity
+
+        dihedral: per-dihedral quantity
+
+        improper: per-improper quantity
+
+        pair: per-pair quantity
+
+        particle: per-particle quantity
+
+        state: internal flag for specifying object's internal state
+
+        ALL: a combination of all other flags
+
+        NONE: represents no flag
+    """
+    NONE = 0
+    scalar = auto()
+    sequence = auto()
+    string = auto()
+    strings = auto()
+    object = auto()
+    angle = auto()
+    bond = auto()
+    constraint = auto()
+    dihedral = auto()
+    improper = auto()
+    pair = auto()
+    particle = auto()
+    state = auto()
 
     @classmethod
-    def log(cls, func=None, is_property=True, flag='scalar'):
-        def helper(func):
-            name = func.__name__
-            if name in cls._meta_export_dict:
-                raise KeyError(
-                    "Multiple loggable quantities named {}.".format(name))
-            cls._meta_export_dict[name] = flag
-            if is_property:
-                return property(func)
+    def any(cls, flags):
+        def from_str(flag):
+            if isinstance(flag, str):
+                return cls[flag]
             else:
-                return func
-        if func is None:
-            return helper
-        else:
-            return helper(func)
-
-    def __new__(cls, name, base, dct):
-        new_cls = super().__new__(cls, name, base, dct)
-        log_dict = dict()
-        for name, flag in cls._meta_export_dict.items():
-            log_dict[name] = LoggerQuantity(name, new_cls, flag)
-        if hasattr(new_cls, '_export_dict'):
-            old_dict = deepcopy(new_cls._export_dict)
-            for key, value in old_dict.items():
-                old_dict[key] = value.update_cls(new_cls)
-            old_dict.update(log_dict)
-            new_cls._export_dict = old_dict
+                return flag
+
+        return reduce(
+            lambda x, y: from_str(x) | from_str(y), flags, TypeFlags.NONE)
+
+    @classmethod
+    def _get_string_list(cls, flag):
+        return [mem.name for mem in cls.__members__.values() if mem in flag]
+
+
+TypeFlags.ALL = TypeFlags.any(TypeFlags.__members__.values())
+
+
+def _loggables(self):
+    """dict[str, str] Return a name: flag mapping of loggable quantities
+    for the class."""
+    return {name: quantity.flag.name
+            for name, quantity in self._export_dict.items()}
+
+
+class Loggable(type):
+    _meta_export_dict = dict()
```

Comment:
This is needed as we need a place to store the `_LoggerQuantities` intermediately between the definition and creation of the class. We can't store in in the classes which use `Loggable` as a metaclass since they won't be created to store a `dict` in until after `__new__` is called on `type` to create the class.

That is why we empty the attribute after every call of `__init__`.